### ISSUE TITLE
Tag prefix broken on Windows with versioneer 0.21

### ISSUE THREAD

**Issue by jenshnielsen** [python-versioneer/python-versioneer#273]
https://github.com/python-versioneer/python-versioneer/pull/262 contains a fix for handling * on windows unfortunately this breaks detecting the correct version for me.

In our repository https://github.com/qcodes/qcodes all tags are prefixed with `v` and we have `tag_prefix = v` this used to work fine however with the latest release of versioneer we get versions reported as unknown + x

Looking at the generated `git describe` command we see 

```
git describe --tags --always --long --dirty --match v*
v0.29.0-288-g9f93a1286
```

```
git describe --tags --always --long --dirty --match v\*
9f93a1286
```

```
git describe --tags --always --long --dirty --match v\\*
9f93a1286
```

so it would seem like the escape of * does not always work correctly.
removing the escaped prefix for windows restores normal functionality for us

**Comment by hameerabbasi** [comment]
Is there a known working version of versioneer I can use? I tried [0.21](https://github.com/Quansight-Labs/uarray/runs/4300847003?check_suite_focus=true) and [0.20](https://github.com/Quansight-Labs/uarray/runs/4302280440?check_suite_focus=true) without success, both produced wheels that were untagged on Windows.

**Comment by jenshnielsen** [comment]
So for me the conclusion was the 0.20 works in powershell/cmd while 0.21 works in git bash. I don't think there is any version that works every where

**Comment by hameerabbasi** [comment]
Would there be a reason both 0.21 and 0.20 produce untagged wheels on Windows with git bash? Is that a separate issue?

**Comment by jenshnielsen** [comment]
I would think that is a separate issue. To debug I would suggest that you confirm that the the `git describe function` returns something meaningful if you run it from the git bash command line. You can see here https://github.com/python-versioneer/python-versioneer/blob/master/src/git/from_vcs.py#L39 which command versioneer executes

**Comment by jmartens** [comment]
The same happens using a cmd prompt on the same system:

```
>git describe --tags --dirty --always --long --match \\*
6408ff9

>git describe --tags --dirty --always --long --match *
2022.01.1-0-g6408ff9

C:\Users\jonathan.martens\Code\raystation-administration>venv\Scripts\activate

(venv) C:\Users\jonathan.martens\Code\raystation-administration>python setup.py version
running version
keywords are unexpanded, not using
got version from VCS {'version': '0+untagged.87.g6408ff9', 'full-revisionid': '6408ff9937d1473be330c44a5392a23e23a35113', 'dirty': False, 'error': None, 'date': '2022-01-31T14:41:24+0100'}
Version: 0+untagged.87.g6408ff9
 full-revisionid: 6408ff9937d1473be330c44a5392a23e23a35113
 dirty: False
 date: 2022-01-31T14:41:24+0100

(venv) >
```

**Comment by jmartens** [comment]
I am running the following config in both cases:
```
[versioneer]
VCS = git
style = pep440
versionfile_source = _version.py
tag_prefix =
```

**Comment by effigies** [comment]
@jmartens The issue is that we get different behavior based on whether someone's using CMD/Powershell or bash, which are both common workflows. I made an attempt at resolving this in https://github.com/python-versioneer/python-versioneer/pull/274, but didn't have time to track this down, as I don't have any Windows machines. I would be happy to hand that PR over to someone who can test all these cases.

**Comment by jmartens** [comment]
For me using power shell with an empty prefix this makes it work again:

```
    if sys.platform == "win32":
        GITS = ["git.cmd", "git.exe"]
        if not tag_prefix:
            TAG_PREFIX_REGEX = r"\*"
```

**Comment by jmartens** [comment]
> I made an attempt at resolving this in #274, but didn't have time to track this down, as I don't have any Windows machines.

So that fix does not work 100% either, so I think #274 should be reverted as this also has undesired side effects for other cases as @jenshnielsen reported and I am suffering as well.

**Comment by mathijsvdv** [comment]
> @jmartens The issue is that we get different behavior based on whether someone's using CMD/Powershell or bash, which are both common workflows. I made an attempt at resolving this in #274, but didn't have time to track this down, as I don't have any Windows machines. I would be happy to hand that PR over to someone who can test all these cases.

Today I ran into the same issue (I was using v* tags) on windows where, at least on CMD, I was able to fix it by putting
```
    if sys.platform == "win32":
        GITS = ["git.cmd", "git.exe"]
```
without the `TAG_PREFIX_REGEX`. I'd be happy to look into the other cases as well. From the discussion, there is possibly different behavior to take into account for the following:
- Windows vs Linux
- CMD vs Powershell vs bash
- Prefix or no prefix

I would be happy to pick up this issue as I am working on windows myself and can use WSL where needed for linux testing. What I'd suggest is that we build #263 into a pull request that also takes into account the above cases. Without the CI, we can't be certain that we're covering everything.

How does this sound?
